tests/test_override_loop.py

```diff
@@ -133,6 +133,78 @@ async def blah():
         asyncio.get_event_loop().run_until_complete(info["coro"])
 
 
+it "can shutdown async gens":
+    info1 = []
+    info2 = []
+    info3 = []
+
+    original = asyncio.get_event_loop()
+
+    async def my_generator(info):
+        try:
+            info.append(1)
+            yield
+            info.append(2)
+            yield
+            info.append(3)
+        except asyncio.CancelledError:
+            info.append("cancelled")
+            raise
+        finally:
+            info.append(("done", __import__("sys").exc_info()[0]))
+
+    # Test that the outside loop isn't affected by the inside loop
+    outside_gen = my_generator(info1)
+
+    async def outside1():
+        await outside_gen.__anext__()
+        await outside_gen.__anext__()
+
+    original.run_until_complete(outside1())
+    assert info1 == [1, 2]
+
+    # The way python asyncio works
+    # Means that by defining this outside our OverrideLoop
+    # The weakref held against it in the _asyncgens set on the loop
+    # Will remain so that our shutdown_asyncgens function may work
+    ag = my_generator(info2)
+
+    with OverrideLoop(new_loop=True) as custom_loop:
+        assert info2 == []
+        assert info3 == []
+
+        async def doit():
+            ag2 = my_generator(info3)
+            assert set(asyncio.get_event_loop()._asyncgens) == set()
+            await ag2.__anext__()
+            assert set(asyncio.get_event_loop()._asyncgens) == set([ag2])
+            await ag.__anext__()
+            assert set(asyncio.get_event_loop()._asyncgens) == set([ag2, ag])
+            await ag.__anext__()
+            assert info3 == [1]
+
+        custom_loop.run_until_complete(doit())
+        assert list(custom_loop.loop._asyncgens) == [ag]
```

Comment:
This depends on the implementation of the GC, because it assumes `ag2` is collected before this line. That's probably fine given how CPython's GC works right now, but it's not guaranteed to stay the same, so this test might break in a future version of Python (or even current implementations of Python other than CPython).